### PR TITLE
Dependency Update

### PR DESCRIPTION
**Pull request by Superhepper**
Updates all the dependencies of the crate in order to be able to use newer versions of rust and and
remove old lints and checks.

- Updates Rust edition to 2021
- Updates MSRV to 1.74.0
- Updates crate dependencies.
- Updates lock file.
- Updates minimum supported tpm2-tss version to 4.0.1.

Fixes https://github.com/parallaxsecond/rust-tss-esapi/issues/516
Closes https://github.com/parallaxsecond/rust-tss-esapi/pull/540

### CODE REVIEW

File: tss-esapi-sys/README.md [hunk 1]
```diff
@@ -18,7 +18,7 @@ must therefore link to all of them at build time.
 
 The paths to the libraries are discovered using `pkg-config` - make sure they
 are discoverable in this way on your system. Our build script looks for 
-`tss2-esys`, `tss2-tctildr` and `tss2-mu`. A minimum version of `3.2.2` is 
+`tss2-esys`, `tss2-tctildr` and `tss2-mu`. A minimum version of `4.0.1` is
```
Comment (ionut-arm) [inline tss-esapi-sys/README.md:21]: Should we change [the version we get in our test containers](https://github.com/parallaxsecond/rust-tss-esapi/blob/main/tss-esapi/tests/Dockerfile-ubuntu#L10)?

**Review by ionut-arm — COMMENTED** [review]
Amazing stuff, thanks for the patch, I'm in favour of these changes!

What's missing? Why is it a draft?

**Comment by Superhepper** [comment]
> Amazing stuff, thanks for the patch, I'm in favour of these changes!
> 
> What's missing? Why is it a draft?

I wanted peoples input before making it a real PR.

File: tss-esapi-sys/README.md (hunk 1 above)
Comment (Superhepper) [reply tss-esapi-sys/README.md:21]: Yes, I realized I have missed updating the version in other places as well.

Comment (Superhepper) [reply tss-esapi-sys/README.md:21]: But I am having some issues getting the ```regenerate_bindings``` script to work with  4.0.1

Comment (Superhepper) [reply tss-esapi-sys/README.md:21]: Fixed now

File: tss-esapi/tests/Dockerfile-ubuntu [hunk 2]
```diff
@@ -7,10 +7,11 @@ ENV PATH="/root/.cargo/bin:${PATH}"
 
 FROM rust-toolchain AS tpm2-tss
 # Download and install the TSS library
-ENV TPM2_TSS_BINDINGS_VERSION=3.2.2
+ENV TPM2_TSS_BINDINGS_VERSION=4.0.1
 ARG TPM2_TSS_VERSION=$TPM2_TSS_BINDINGS_VERSION
 ENV TPM2_TSS_VERSION=$TPM2_TSS_VERSION
-ENV PKG_CONFIG_PATH /usr/local/lib/pkgconfig
+ENV PKG_CONFIG_PATH=/usr/local/lib/pkgconfig
```
Comment (wiktor-k) [inline tss-esapi/tests/Dockerfile-ubuntu:13]: Why remove the equals sign here while it's there a couple of lines above? :thinking:

**Review by wiktor-k — APPROVED** [review]
Quite a lot of changes. Looks good, thanks! :+1:

**Comment by wiktor-k** [comment]
By the way I've edited this PR description to add "fixes" and "closes". See if this accurate (I like it when the issues/PRs are cross-linked). :wave:

File: tss-esapi/tests/Dockerfile-ubuntu (hunk 2 above)
Comment (Superhepper) [reply tss-esapi/tests/Dockerfile-ubuntu:13]: I think I got a warning when I was building the docker about deprecated initialization of variables or something like that. And that is why I changed it.

File: tss-esapi-sys/regenerate-bindings.sh [hunk 3]
```diff
@@ -48,7 +48,7 @@ cross-compile-openssl() {
 cross-compile-tpm2-tss() {
     pushd /tmp/tpm2-tss
     [ ! -f configure ] && ./bootstrap
-    ./configure --enable-fapi=no --prefix=/ --build=x86_64-pc-linux-gnu --host=$1 --target=$1 CC=$1-gcc
+    ./configure --disable-fapi --disable-policy --prefix=/ --build=x86_64-pc-linux-gnu --host=$1 --target=$1 CC=$1-gcc
```
Comment (ionut-arm) [inline tss-esapi-sys/regenerate-bindings.sh:51]: Hmm, I kinda hope this doesn't disable policies for ESAPI as well... 🤔

**Review by ionut-arm — APPROVED** [review]
Looks good, just one comment

File: tss-esapi-sys/regenerate-bindings.sh (hunk 3 above)
Comment (Superhepper) [reply tss-esapi-sys/regenerate-bindings.sh:51]: I tried to look deep in their config files to exact what the flag did but got lost in how much I do not ... approve .. of autoconf files and syntax. But I saw that the Yocto project made some similair changes in a patch a looked for mocing to 4.0.1.

**Pull request merged**